### PR TITLE
refactor: Manage async tasks with task manager

### PR DESCRIPTION
**Pull request by Reto911**
In this PR, we managed asynchronous tasks with the `TaskManager` introduced in #478.

### CODE REVIEW

File: src/common/task_manager/gc.rs [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+
+    /// Checks if the task manager is shutdown.
+    #[inline]
+    #[must_use]
+    pub fn is_shutdown(&self) -> bool {
+        self.status.load(Ordering::Acquire)
+    }
+
+    /// Spawn a task, and send its handle to GC task for managing. Returns the
+    /// abort handle.
```
Comment (Reto911) [inline src/common/task_manager/gc.rs:61]: Fix the comment.

File: src/storage/memory_cache/write_back_task.rs [hunk 2]
```diff
@@ -112,15 +141,34 @@ where
...
+                            tx,
+                        )
+                    })
+                    .await
+                    .is_err()
+                {
+                    // Ensure that the block is flushed to the backend, and notify the loop to
+                    // shutdown.
+                    error!("Try to spawn a `BlockFlush` task after shutdown.");
+                    if let Err(e) = self.storage.backend().store(ino, block_id, block).await {
```
Comment (xiaguan) [inline src/storage/memory_cache/write_back_task.rs:166]: When shutdown ,just return directly.

Comment (Reto911) [reply src/storage/memory_cache/write_back_task.rs:166]: Print the error in log, and ignore the un-flushed block?

File: src/storage/memory_cache/write_back_task.rs [hunk 3]
```diff
@@ -131,18 +179,56 @@ where
...
+        let mut shutdown = self.block_flush_spawn_handle.is_shutdown();
+
         if let Some(file_level_pending_blocks) = file_level_pending_blocks {
-            for (block_id, (block, sender)) in file_level_pending_blocks {
-                let res = self.storage.backend().store(ino, block_id, block).await;
-                sender.send(res).unwrap_or_else(|res| {
-                    error!("The receiver is closed unexpectedly, with storage result: {res:?}");
-                });
+            for (block_id, (block, tx)) in file_level_pending_blocks {
+                if shutdown {
```
Comment (xiaguan) [inline src/storage/memory_cache/write_back_task.rs:187]: When a user initiates a logout, I believe we only need to ensure the program exits normally. Try not to do anything extra as much as possible

File: src/common/task_manager/manager.rs [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+#[error("Failed to spawn task: {0:?}")]
+pub struct SpawnError(pub TaskName);
+
+/// The task manager, which will shutdown all async tasks in proper order.
+#[derive(Debug)]
+pub struct TaskManager {
+    /// Tasks in task manager.
+    tasks: Mutex<HashMap<TaskName, Task>>,
+    /// The status of task manager, `true` for shutting down.
+    status: Arc<AtomicBool>,
```
Comment (xiaguan) [inline src/common/task_manager/manager.rs:33]: rename to `is_shutdown`

**Review by xiaguan — CHANGES_REQUESTED** [review]

File: src/common/task_manager/manager.rs [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+    let future = async move {
+        if let Some(signal) = signals.next().await {
+            debug_assert!(
+                TERM_SIGNALS.contains(&signal),
+                "The signal hook is not to handle signal {signal}."
+            );
+            info!("Signal {signal} raised, start to shutdown.");
+        } else {
+            info!("The signal stream is closed.");
```
Comment (xiaguan) [inline src/common/task_manager/manager.rs:205]: I think it's a error.

File: src/common/task_manager/manager.rs [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+    let mut signals = Signals::new(TERM_SIGNALS)?;
+    let handle = signals.handle();
+
+    let future = async move {
+        if let Some(signal) = signals.next().await {
+            debug_assert!(
+                TERM_SIGNALS.contains(&signal),
+                "The signal hook is not to handle signal {signal}."
+            );
+            info!("Signal {signal} raised, start to shutdown.");
```
Comment (xiaguan) [inline src/common/task_manager/manager.rs:203]: It's a warn.

File: src/common/task_manager/manager.rs [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    pub async fn shutdown(&self) {
+        let mut queue = VecDeque::from([TaskName::Root]);
+
+        self.status.store(true, Ordering::Release);
+
+        let mut tasks = std::mem::take(&mut *self.tasks.lock().await);
+
+        while let Some(task_name) = queue.pop_front() {
+            let Some(mut task_node) = tasks.remove(&task_name) else {
+                continue;
```
Comment (xiaguan) [inline src/common/task_manager/manager.rs:153]: Need log here.

File: src/common/task_manager/manager.rs (hunk 6 above)
Comment (Reto911) [reply src/common/task_manager/manager.rs:203]: I don't think it's a warn. A signal to shutdown is such a normal situation, such as `Ctrl+C`, `kill` and `docker stop`.

**Review by xiaguan — APPROVED** [review]

**Pull request merged**